utility/map_str_enhancer.go

```diff
@@ -4,44 +4,99 @@ import (
 	"github.com/elastic/beats/libbeat/common"
 )
 
-func AddStrWithDefault(m common.MapStr, key string, val *string, defaultVal string) {
-	if val != nil {
-		m[key] = *val
-	} else if defaultVal != "" {
-		m[key] = defaultVal
-	}
-}
-
 func Add(m common.MapStr, key string, val interface{}) {
-
+	if m == nil || key == "" {
+		return
+	}
 	switch val.(type) {
 	case *bool:
 		if newVal := val.(*bool); newVal != nil {
 			m[key] = *newVal
+		} else {
+			delete(m, key)
 		}
 	case *int:
 		if newVal := val.(*int); newVal != nil {
 			m[key] = *newVal
+		} else {
+			delete(m, key)
 		}
 	case *string:
 		if newVal := val.(*string); newVal != nil {
 			m[key] = *newVal
+		} else {
+			delete(m, key)
 		}
 	case common.MapStr:
 		if valMap := val.(common.MapStr); len(valMap) > 0 {
+			for k, v := range valMap {
+				Add(valMap, k, v)
+			}
 			m[key] = valMap
+			if len(valMap) > 0 {
+				m[key] = valMap
+			} else {
+				delete(m, key)
+			}
+		} else {
+			delete(m, key)
+		}
+	case map[string]interface{}:
+		valMap := val.(map[string]interface{})
```

Comment:
this wouldn't handle something like `{"a: {"b": null}}`, right?

Reply:
It should, as I iterate through all the elements (see below). 
I added a spec to show this case though.

Reply:
ah ofc 🤦‍♂️ sorry